### DISCUSSION TITLE
Is it possible to split routing into multiple files with the new data APIs?

### DISCUSSION THREAD

**Discussion by WonkiestNebula** [remix-run/react-router#10160] [Q&A] [answered]
In the older `<BrowserRouter />` style, you could have several files that split the burden of routing for larger apps. I was wanting to use some of the new Data API features, but I can't seem to find any way to do that. Am I missing something, or is that not an option at this point?  
  
Example of the old style:
```js
//RouterOne.jsx
const RouterOne = () => {
  return (
    <Routes>
      <Route path='pathOne/*' element={<RouterTwo />} />
    </Routes>
  )
}
```
```js
//RouterTwo.jsx
const RouterTwo = () => {
  return (
    <Routes>
      <Route path='pathTwo' element={<OtherThing />} />
    </Routes>
  )
}
```
  
And `<OtherThing />` would be accessible at the route `/pathOne/pathTwo`

**Comment by dead1ock** [comment, marked as answer]
I haven't found a way to do this using `createRoutesFromElements()` without an error about duplicate `routeId`s. This is how I approached splitting up Routes using module exports. 

```js
// src/App.tsx
import { createBrowserRouter, RouterProvider } from 'react-router-dom';
import AppRoutes from 'routes';

const router = createBrowserRouter(AppRoutes);

export const App = () => {
  return (
      <RouterProvider router={router} />
  );
};
```

```js
// src/routes/index.tsx

import AuthRoutes from './auth';
import ProfileRoute from './profile';

const AppRoutes = [
  {
    path: '/',
    element: <Root />
    children: [
      AuthRoutes,
      ProfileRoute,
      // etc...
    ]
];

export default AppRoutes;
```

```js
// src/routes/auth/index.tsx

import LoginRoute from './login';
import LogoutRoute from './logout';
import CallbackRoute from './callback';

const AuthRoutes = {
  path: 'auth',
  children: [
    LoginRoute,
    LogoutRoute,
    CallbackRoute
  ]
}

export default AuthRoutes;
```

```js
// src/routes/auth/login/index.tsx

import LoginPage from './LoginPage';

const LoginRoute = {
  path: 'login',
  element: <LoginPage />
}

export default LoginRoute;
```

```js
// src/routes/profile/index.tsx

import ProfilePage from './ProfilePage';
import loader from './loader';

const ProfileRoute = {
  path: 'profile',
  element: <ProfilePage />
  loader
}

export default ProfileRoute;
```

**Reply by brophdawg11** [reply]
Yeah at the end of the day routes are just JSON objects so you're free to compose them together any way you like (as shown above).  The JSX notation allowed you to dynamically define/remove routes via the render lifecycle in `BrowserRouter` apps.  But in `RouterProvider` the JSX is there as a convenience (and it can be easier on the eyes) but they just get converted to JSON in `createRoutesFromElements`.

If you're relying on `createRoutesFromElements` to generate your route IDs, then yeah you'd need a single root-level call so it can properly generate globally unique route IDs.  There's an example of how you might achieve that in https://github.com/remix-run/react-router/discussions/10373#discussioncomment-5676547.  

But if you are providing your own route IDs then you should be able to compose multiple `createRoutesFromElements` calls without issue.

**Reply by jasikpark** [reply]
Is there a way to mix and match "createRoutesFromElements" so I could slowly transition from the JSX design to the JSON design? like

```tsx
import routeObject from "./LolRoot"
<Route path="lol/*" {...routeObject}/>
```

basically a way to make a JSON route object be consumed by a JSX Route to allow the same sort of "walk up the nodes" transition that `react-router-dom-v5-compat` allows?

**Reply by jasikpark** [reply]
ig the answer would be to move the other direction...

**Reply by brophdawg11** [reply]
Yes the spread syntax should work fine.  I would just move to JSON route by route as you lift a route from `<Routes>` to the `createBrowserRouter`.

**Reply by jasikpark** [reply]
A more concrete example:

```tsx

export const GetStartedRootRouteConfig = {
  element: <Titled title={() => page('Get started')} />,
  children: [
    { path: 'revisit', element: <RevisitGettingStarted /> },
    {
      index: true,
      Component: () => {
        const navigate = useNavigate();
        return (
          <GetStartedPage
            onSuccess={() => navigate('/lighthouses', { replace: true })}
            onRevisit={() => navigate('revisit', { replace: true })}
          />
        );
      },
    },
  ],
} satisfies RouteObject;
```

```tsx
<Route path="get-started/*" {...GetStartedRootRouteConfig} />
```

doesn't typecheck w/ the error

```
Type '{ element: Element; children: ({ path: string; element: Element; index?: undefined; Component?: undefined; } | { index: true; Component: () => Element; path?: undefined; element?: undefined; })[]; path: string; }' is not assignable to type 'IntrinsicAttributes & RouteProps'.
  Property 'index' is missing in type '{ element: Element; children: ({ path: string; element: Element; index?: undefined; Component?: undefined; } | { index: true; Component: () => Element; path?: undefined; element?: undefined; })[]; path: string; }' but required in type 'IndexRouteProps'.
```

**Reply by jasikpark** [reply]
<img width="489" alt="image" src="https://user-images.githubusercontent.com/10626596/233503641-a8e59859-90d2-4c6a-852a-1f64b8fa0598.png">
ah, additional problem with spreading the nested routes in my example

**Reply by jasikpark** [reply]
So it seems I need to start at the routing tree root, and work my way out to the leaves

**Reply by AprilMay812** [reply]
Oh my god, thank you so much. I'm relatively new to using react-router, and you have no idea how much of a headache this was giving me trying to find an answer to this. Maybe I just need to work on my google-fu lol, but I could not for the life of me find an answer to this anywhere, or at least not one that made sense to me....

**Comment by amangeot** [comment]
Hello,
This seems to be working using a hook, even though passing a component returning the same JSX will throw an error:

```jsx
export const useRoutes = () => {
  return (
    <>
      <Route path='/path' element={<Element />} />
      <Route path='/some-other-path' element={<Element />} />
    </>
  );
};
```

along with:
```jsx
export const MyRoutes = () => {
  const routes = useRoutes();

  return (
    <Routes>
      {routes}
      <Route path="*" element={<NotFound />} />
    </Routes>
  );
};
```

**Reply by jch422** [reply]
don't know why but my code works with

```jsx
<Routes>
  {routes()}
  <Route path="*" element={<NotFound />} />
</Routes>
```
